indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java

```diff
@@ -1018,4 +1022,37 @@ List<Task> getTasks()
       giant.unlock();
     }
   }
+
+  void validateTaskPayload(Task task)
+  {
+    try {
+      String payload = passwordRedactingMapper.writeValueAsString(task);
+      if (payload.length() > TASK_SIZE_WARNING_THRESHOLD) {
+        log.warn("Received a large task payload [%s] with id [%s] and datasource [%s]" +
+                " There may be downstream issues caused by managing this large payload." +
+                "Set druid.indexer.queue.maxTaskPayloadSize to reject tasks above a certain size.",
+            payload.length(),
+            task.getId(),
+            task.getDataSource()
+        );
+      }
+
+      if (config.getMaxTaskPayloadSize() != null && config.getMaxTaskPayloadSize().getBytesInInt() < payload.length()) {
+        throw DruidException.forPersona(DruidException.Persona.OPERATOR)
+            .ofCategory(DruidException.Category.INVALID_INPUT)
+            .build(
+                "Task payload size was [%d] but max size is [%d]. " +
+                    "Reduce the size of the task or increase 'druid.indexer.queue.maxTaskPayloadSize'.",
+                payload.length(),
+                config.getMaxTaskPayloadSize()
+            );
```

Comment:
Rephrased, used short-hand for invalid input exception, included task id in error message:

```suggestion
        throw InvalidInput.exception(
                "Task[%s] has payload of size[%d] but max allowed size is [%d]. " +
                    "Reduce the size of the task payload or increase 'druid.indexer.queue.maxTaskPayloadSize'.",
                task.getId(), payload.length(), config.getMaxTaskPayloadSize()
            );
```